File: LR1/src/main/scala/comparator/FractionComparator.scala
package comparator

import types.FractionType

class FractionComparator extends Comparator with Serializable {
  override def compare(o1: AnyRef, o2: AnyRef): Int = {
    val lcm = FractionType.lcm(o1.asInstanceOf[FractionType].getDenominatorValue, o2.asInstanceOf[FractionType].getDenominatorValue)
    val coef1 = if (o1.asInstanceOf[FractionType].getIsNegative) -1
    else 1
    val coef2 = if (o2.asInstanceOf[FractionType].getIsNegative) -1
    else 1
    val num1 = coef1 * (o1.asInstanceOf[FractionType].getIntegerPartValue * o1.asInstanceOf[FractionType].getDenominatorValue + o1.asInstanceOf[FractionType].getNumeratorValue) * (lcm / o1.asInstanceOf[FractionType].getDenominatorValue)
    val num2 = coef2 * (o2.asInstanceOf[FractionType].getIntegerPartValue * o2.asInstanceOf[FractionType].getDenominatorValue + o2.asInstanceOf[FractionType].getNumeratorValue) * (lcm / o2.asInstanceOf[FractionType].getDenominatorValue)
    Integer.compare(num1, num2)
  }
}

File: LR3/app/src/main/java/com/example/lr3/comparator/FractionComparator.java
package com.example.lr3.comparator;

import com.example.lr3.types.FractionType;

import java.io.Serializable;

public class FractionComparator implements Comparator, Serializable {
    @Override
    public int compare(Object o1, Object o2) {
        Integer lcm = FractionType.lcm(((FractionType) o1).getDenominatorValue(), ((FractionType) o2).getDenominatorValue());
        int coef1 = ((FractionType) o1).getIsNegative() ? -1 : 1;
        int coef2 = ((FractionType) o2).getIsNegative() ? -1 : 1;

        int num1 = coef1 * (((FractionType) o1).getIntegerPartValue() * ((FractionType) o1).getDenominatorValue() +
                ((FractionType) o1).getNumeratorValue()) * (lcm / ((FractionType) o1).getDenominatorValue());
        int num2 = coef2 * (((FractionType) o2).getIntegerPartValue() * ((FractionType) o2).getDenominatorValue() +
                ((FractionType) o2).getNumeratorValue()) * (lcm / ((FractionType) o2).getDenominatorValue());

        return Integer.compare(num1, num2);
    }

}
